### PR TITLE
Ansi colors

### PR DESCRIPTION
**Pull request by fmthoma**
Parse ANSI escape codes in the grep output and format the output text accordingly (results list and pager).

`--color=always` now switched on by default for `grep` invocation.

### CODE REVIEW

File: src/Vgrep/Widget/Results.hs [hunk 1]
```diff
@@ -187,10 +188,9 @@ renderLine width lineNumberWidth displayLine = do
                           . justifyRight lineNumberWidth
                           . maybe "" (T.pack . show)
```
Comment (fmthoma) [inline src/Vgrep/Widget/Results.hs]: Might be worth a PR to `text`...

File: src/Vgrep/Widget/Pager.hs [hunk 2]
```diff
@@ -17,15 +17,15 @@ module Vgrep.Widget.Pager (
...
 import           Data.Foldable
+import qualified Data.IntMap.Strict   as Map
 import           Data.Sequence        (Seq, (><))
 import qualified Data.Sequence        as Seq
-import qualified Data.Set             as Set
 import           Data.Text            (Text)
-import qualified Data.Text            as T
 import           Graphics.Vty.Image   hiding (resize)
 import           Graphics.Vty.Input
 import           Graphics.Vty.Prelude
```
Comment (fmthoma) [inline src/Vgrep/Widget/Pager.hs]: Actually, that's what I use. But I don't currently group qualified and unqualified imports separately.

File: src/Vgrep/Ansi/Parser.hs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    21 -> withoutStyle Vty.bold
+    22 -> withoutStyle Vty.bold
+    23 -> withoutStyle Vty.standout
+    24 -> withoutStyle Vty.underline
+    25 -> withoutStyle Vty.blink
+    27 -> withoutStyle Vty.reverseVideo
+    i | i >= 30  && i <= 37   -> withForeColor (rawColor (i - 30))
+      | i >= 40  && i <= 47   -> withBackColor (rawColor (i - 40))
+      | i >= 90  && i <= 97   -> withForeColor (rawBrightColor (i - 90))
+      | i >= 100 && i <= 107  -> withBackColor (rawBrightColor (i - 100))
```
Comment (fmthoma) [inline src/Vgrep/Ansi/Parser.hs]: Yes, I like to add a bit of visual separation between the guard and the implementation.

File: src/Vgrep/Widget/Pager/Internal.hs [hunk 4]
```diff
@@ -24,7 +26,7 @@ data Pager = Pager
     { _column      :: Int
     -- ^ The current column offset for horizontal scrolling
 
-    , _highlighted :: Set Int
+    , _highlighted :: IntMap (Formatted Attr)
     -- ^ Set of line numbers that are highlighted (i.e. they contain matches)
 
     , _above       :: Seq Text
```
Comment (fmthoma) [inline src/Vgrep/Widget/Pager/Internal.hs]: I think it's not even enabled. I just kept the alignment when adding Haddocks...

Comment (fmthoma) [inline src/Vgrep/Widget/Pager/Internal.hs]: Will change that on master, does not belong in this PR...

File: app/Main.hs [hunk 5]
```diff
@@ -133,11 +133,11 @@ eventHandler = \case
...
         -> Next (VgrepT AppState m Redraw)
-    handleFeedResult line = Continue $ do
-        expandedLine <- expandLineForDisplay line
-        case parseLine expandedLine of
-            Just l  -> zoom results (feedResult l)
-            Nothing -> pure Unchanged
+    handleFeedResult line = Continue $ case parseLine line of
+        Just l  -> do
+            l' <- traverseOf (lineReference . lineText) expandFormattedLine l
+            zoom results (feedResult l')
```
Comment (fmthoma) [inline app/Main.hs]: Apart from the fact that `traverseOf` is kind of ill-named, I think it is pretty clear what this does. `traverseOf` is the monadic variant of `over` that takes a monadic action and applies it to all traversed elements, just like `over` does with a pure function.

And, interestingly enough, `traverseOf = id`, so we could even omit the `traverseOf` here.

File: app/Main.hs [hunk 6]
```diff
@@ -210,12 +210,12 @@ whenJust item action = maybe (pure mempty) action item
 
 invokeEditor :: AppState -> Next (VgrepT AppState m Redraw)
 invokeEditor state = case views (results . currentFileName) (fmap T.unpack) state of
-    Just file -> Interrupt $ Suspend $ \environment -> do
+    Just selectedFile -> Interrupt $ Suspend $ \environment -> do
```
Comment (fmthoma) [inline app/Main.hs]: Does not do any harm, some people even prefer `f $ g $ h $ a` to `f . g . h $ a`.

File: src/Vgrep/Ansi/Parser.hs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    5  -> withStyle Vty.blink
+    6  -> withStyle Vty.blink
+    7  -> withStyle Vty.reverseVideo
+    21 -> withoutStyle Vty.bold
+    22 -> withoutStyle Vty.bold
+    23 -> withoutStyle Vty.standout
+    24 -> withoutStyle Vty.underline
+    25 -> withoutStyle Vty.blink
+    27 -> withoutStyle Vty.reverseVideo
+    i | i >= 30  && i <= 37   -> withForeColor (rawColor (i - 30))
```
Comment (fmthoma) [inline src/Vgrep/Ansi/Parser.hs]: No big difference IMHO. Not worth the extra definition, which has to be read anyway. BTW: What's the correct order for the three `Int` arguments of `between`? ;-)

File: app/Main.hs (hunk 6 above)
Comment (quchen) [inline app/Main.hs]: ((((ok))))

**Comment by fmthoma** [comment]
This branch still exhibits a space leak when rendering a long (~1kb) result line.

File: app/Main.hs (hunk 5 above)
Comment (quchen) [inline app/Main.hs]: I think `id` is *much* less easy to read. Just imagine you come along this piece of code and don’t know it’s from Lens! And if that’s unconvincing, remember that `($) = id` as well, up to a fixity declaration.

File: src/Vgrep/Ansi/Parser.hs (hunk 3 above)
Comment (quchen) [inline src/Vgrep/Ansi/Parser.hs]: Given my space-before-closing-parenthesis I don’t have any good arguments to bring up against this, other than »Noooo don’t do that« :-D

File: src/Vgrep/Widget/Results.hs (hunk 1 above)
Comment (quchen) [inline src/Vgrep/Widget/Results.hs]: Will probably be rejected because of terrible performance, I could imagine

File: app/Main.hs (hunk 5 above)
Comment (fmthoma) [inline app/Main.hs]: I was pointing out the fact, not suggesting to use `id` here :-)
